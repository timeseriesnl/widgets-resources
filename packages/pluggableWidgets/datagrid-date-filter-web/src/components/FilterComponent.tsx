import { createElement, ReactElement, useCallback, useEffect, useRef, useState } from "react";
import { FilterSelector } from "./FilterSelector";

import { DefaultFilterEnum } from "../../typings/DatagridDateFilterProps";

import DatePickerComponent from "react-datepicker";
import { DatePicker } from "./DatePicker";

interface FilterComponentProps {
    adjustable: boolean;
    defaultFilter: DefaultFilterEnum;
    defaultValue?: Date;
    dateFormat?: string;
    locale?: string;
    name?: string;
    placeholder?: string;
    screenReaderButtonCaption?: string;
    screenReaderInputCaption?: string;
    tabIndex?: number;
    updateFilters?: (value: Date | null, type: DefaultFilterEnum) => void;
}

export function FilterComponent(props: FilterComponentProps): ReactElement {
    const [type, setType] = useState<DefaultFilterEnum>(props.defaultFilter);
    const [value, setValue] = useState<Date | null>(null);
    const pickerRef = useRef<DatePickerComponent | null>(null);

    useEffect(() => {
        if (props.defaultValue) {
            setValue(props.defaultValue);
        }
    }, [props.defaultValue]);

    useEffect(() => {
        props.updateFilters?.(value, type);
    }, [value, type]);

    const focusInput = useCallback(() => {
        if (pickerRef.current) {
            pickerRef.current.setFocus();
        }
    }, [pickerRef.current]);

    return (
        <div className="filter-container" data-focusindex={props.tabIndex ?? 0}>
            {props.adjustable && (
                <FilterSelector
                    ariaLabel={props.screenReaderButtonCaption}
                    defaultFilter={props.defaultFilter}
                    name={props.name}
                    onChange={type => {
                        setType(type);
                        focusInput();
                    }}
                />
            )}
            <DatePicker
                adjustable={props.adjustable}
                dateFormat={props.dateFormat}
                locale={props.locale}
                name={props.name}
                placeholder={props.placeholder}
                ref={pickerRef}
                screenReaderInputCaption={props.screenReaderInputCaption}
                setValue={setValue}
                value={value}
            />
        </div>
    );
}
