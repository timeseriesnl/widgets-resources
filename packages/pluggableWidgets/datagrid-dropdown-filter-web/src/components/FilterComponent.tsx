import { createElement, Fragment, ReactElement, useCallback, useEffect, useMemo, useRef, useState } from "react";
import { ListAttributeValue } from "mendix";
import { useOnClickOutside } from "@mendix/piw-utils-internal";
import classNames from "classnames";

export interface FilterOption {
    caption: string;
    value: string;
}

interface FilterComponentProps {
    ariaLabel?: string;
    attribute?: ListAttributeValue;
    auto?: boolean;
    emptyOptionCaption?: string;
    multiSelect?: boolean;
    name?: string;
    options: FilterOption[];
    tabIndex?: number;
    defaultValue?: string;
    updateFilters?: (values: FilterOption[]) => void;
}

export function FilterComponent(props: FilterComponentProps): ReactElement {
    const [valueInput, setValueInput] = useState("");
    const [options, setOptions] = useState<FilterOption[]>([]);
    const [selectedFilters, setSelectedFilters] = useState<FilterOption[]>([]);
    const [show, setShow] = useState(false);
    const [dropdownWidth, setDropdownWidth] = useState(0);

    const componentRef = useRef<HTMLDivElement>(null);

    const setMultiSelectFilters = useCallback(
        (selectedOptions: FilterOption[]) => {
            if (selectedOptions?.length === 0) {
                setValueInput(props.emptyOptionCaption ?? "");
                setSelectedFilters([]);
            } else {
                setValueInput(selectedOptions.map(option => option.caption).join(","));
                setSelectedFilters(selectedOptions);
            }
        },
        [props.emptyOptionCaption]
    );

    const onClick = useCallback(
        (option: FilterOption) => {
            if (props.multiSelect) {
                setMultiSelectFilters(toggleFilter(selectedFilters, option));
            } else {
                setValueInput(option.caption);
                setSelectedFilters([option]);
                setShow(false);
            }
        },
        [selectedFilters, props.emptyOptionCaption, props.multiSelect]
    );

    useOnClickOutside(componentRef, () => setShow(false));

    const availableOptions = useMemo(
        () =>
            props.auto
                ? props.attribute?.universe?.map(value => ({
                      caption: props.attribute?.formatter.format(value) ?? "",
                      value: value?.toString() ?? ""
                  })) ?? []
                : props.options,
        [props.auto, props.attribute, props.options]
    );

    // Select the first option Or default option on load
    useEffect(() => {
        if (props.multiSelect) {
            if (props.defaultValue) {
                const initialOptions = props.defaultValue
                    .split(",")
                    .map(value => availableOptions.find(option => option.value === value))
                    .filter(Boolean) as FilterOption[];

                // User can set anything, but it could not match so we have to set to empty or ""
                setMultiSelectFilters(initialOptions);
            } else {
                setValueInput(props.emptyOptionCaption ?? "");
            }

            setOptions(availableOptions);
        } else {
            // We want to add empty option caption
            const optionsWithEmptyCaption = [
                { caption: props.emptyOptionCaption ?? "", value: "" },
                ...availableOptions
            ];
            const initialOption =
                optionsWithEmptyCaption.find(option => option.value === props.defaultValue) ??
                optionsWithEmptyCaption[0];

            setValueInput(initialOption?.caption ?? "");
            setSelectedFilters([initialOption]);
            setOptions(optionsWithEmptyCaption);
        }
    }, [props.defaultValue, availableOptions, props.emptyOptionCaption]);

    useEffect(() => {
        props.updateFilters?.(selectedFilters);
    }, [selectedFilters]);

    const showPlaceholder = selectedFilters.length === 0 || valueInput === props.emptyOptionCaption;

    return (
        <div className="dropdown-container" data-focusindex={props.tabIndex ?? 0} ref={componentRef}>
            <input
                value={!showPlaceholder ? valueInput : ""}
                placeholder={showPlaceholder ? props.emptyOptionCaption : undefined}
                className="form-control dropdown-triggerer"
                onClick={() => setShow(true)}
                onFocus={() => setShow(true)}
                aria-haspopup
                ref={inputRef => {
                    if (inputRef && inputRef.clientWidth) {
                        setDropdownWidth(inputRef.clientWidth);
                    }
                }}
                aria-expanded={show}
                aria-controls={`${props.name}-dropdown-list`}
                aria-label={props.ariaLabel}
            />
            {show && (
                <ul
                    id={`${props.name}-dropdown-list`}
                    className="dropdown-list"
                    style={{ width: dropdownWidth }}
                    role="menu"
                    data-focusindex={0}
                >
                    {options.map((option, index) => (
                        <li
                            className={classNames({
                                "filter-selected": !props.multiSelect && selectedFilters.includes(option)
                            })}
                            key={index}
                            onClick={() => onClick(option)}
                            onKeyDown={e => {
                                if (e.key === "Enter" || e.key === " ") {
                                    e.preventDefault();
                                    onClick(option);
                                }
                            }}
                            role="menuitem"
                            tabIndex={0}
                        >
                            {props.multiSelect ? (
                                <Fragment>
                                    <input
                                        id={`checkbox_toggle_${index}`}
                                        type="checkbox"
                                        checked={selectedFilters.includes(option)}
                                    />
                                    <label htmlFor={`checkbox_toggle_${index}`} style={{ pointerEvents: "none" }}>
                                        {option.caption}
                                    </label>
                                </Fragment>
                            ) : (
                                <div className="filter-label">{option.caption}</div>
                            )}
                        </li>
                    ))}
                </ul>
            )}
        </div>
    );
}

function toggleFilter(filters: FilterOption[], filterToToggle: FilterOption): FilterOption[] {
    const alteredFilters = [...filters];
    const index = filters.indexOf(filterToToggle);
    if (index > -1) {
        alteredFilters.splice(index, 1);
    } else {
        alteredFilters.push(filterToToggle);
    }

    return alteredFilters;
}
