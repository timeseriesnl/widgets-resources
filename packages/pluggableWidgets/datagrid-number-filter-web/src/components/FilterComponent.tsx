import { createElement, ReactElement, useCallback, useEffect, useRef, useState } from "react";
import { FilterSelector } from "./FilterSelector";

import { DefaultFilterEnum } from "../../typings/DatagridNumberFilterProps";
import { debounce } from "../utils/utils";
import { Big } from "big.js";
import classNames from "classnames";

interface FilterComponentProps {
    adjustable: boolean;
    defaultFilter: DefaultFilterEnum;
    delay: number;
    name?: string;
    placeholder?: string;
    screenReaderButtonCaption?: string;
    screenReaderInputCaption?: string;
    tabIndex?: number;
    updateFilters?: (value: Big | undefined, type: DefaultFilterEnum) => void;
    value?: Big;
}

export function FilterComponent(props: FilterComponentProps): ReactElement {
    const [type, setType] = useState<DefaultFilterEnum>(props.defaultFilter);
    const [value, setValue] = useState<Big | undefined>(undefined);
    const [valueInput, setValueInput] = useState<string | undefined>(undefined);
    const inputRef = useRef<HTMLInputElement | null>(null);

    useEffect(() => {
        if (props.value) {
            setValueInput(props.value.toString());
            setValue(props.value);
        }
    }, [props.value]);

    useEffect(() => {
        props.updateFilters?.(value, type);
    }, [value, type]);

    const onChange = useCallback(
        debounce((value?: Big) => setValue(value), props.delay),
        [props.delay]
    );

    const focusInput = useCallback(() => {
        if (inputRef.current) {
            inputRef.current.focus();
        }
    }, [inputRef]);

    return (
        <div className="filter-container" data-focusindex={props.tabIndex ?? 0}>
            {props.adjustable && (
                <FilterSelector
                    ariaLabel={props.screenReaderButtonCaption}
                    name={props.name}
                    defaultFilter={props.defaultFilter}
                    onChange={type => {
                        setType(type);
                        focusInput();
                    }}
                />
            )}
            <input
                aria-label={props.screenReaderInputCaption}
                className={classNames("form-control", { "filter-input": props.adjustable })}
                onChange={e => {
                    const value = e.target.value;
                    if (value && !isNaN(Number(value))) {
                        setValueInput(value);
                        onChange(new Big(Number(value)));
                    } else {
                        setValueInput(value);
                        onChange(undefined);
                    }
                }}
                placeholder={props.placeholder}
                ref={inputRef}
                type="number"
                value={valueInput}
            />
        </div>
    );
}
