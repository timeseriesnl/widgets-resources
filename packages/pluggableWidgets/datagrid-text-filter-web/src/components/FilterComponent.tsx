import { createElement, ReactElement, useCallback, useEffect, useRef, useState } from "react";
import { FilterSelector } from "./FilterSelector";

import { DefaultFilterEnum } from "../../typings/DatagridTextFilterProps";
import { debounce } from "../utils/utils";
import classNames from "classnames";

interface FilterComponentProps {
    adjustable: boolean;
    defaultFilter: DefaultFilterEnum;
    delay: number;
    name?: string;
    placeholder?: string;
    tabIndex?: number;
    screenReaderButtonCaption?: string;
    screenReaderInputCaption?: string;
    updateFilters?: (value: string, type: DefaultFilterEnum) => void;
    value?: string;
}

export function FilterComponent(props: FilterComponentProps): ReactElement {
    const [type, setType] = useState<DefaultFilterEnum>(props.defaultFilter);
    const [value, setValue] = useState("");
    const [valueInput, setValueInput] = useState("");
    const inputRef = useRef<HTMLInputElement | null>(null);

    useEffect(() => {
        if (props.value) {
            setValueInput(props.value);
            setValue(props.value);
        }
    }, [props.value]);

    useEffect(() => {
        props.updateFilters?.(value, type);
    }, [value, type]);

    const onChange = useCallback(
        debounce((value: string) => setValue(value), props.delay),
        [props.delay]
    );

    const focusInput = useCallback(() => {
        if (inputRef.current) {
            inputRef.current.focus();
        }
    }, [inputRef]);

    return (
        <div className="filter-container" data-focusindex={props.tabIndex ?? 0}>
            {props.adjustable && (
                <FilterSelector
                    ariaLabel={props.screenReaderButtonCaption}
                    name={props.name}
                    defaultFilter={props.defaultFilter}
                    onChange={type => {
                        setType(type);
                        focusInput();
                    }}
                />
            )}
            <input
                aria-label={props.screenReaderInputCaption}
                className={classNames("form-control", { "filter-input": props.adjustable })}
                onChange={e => {
                    setValueInput(e.target.value);
                    onChange(e.target.value);
                }}
                placeholder={props.placeholder}
                ref={inputRef}
                type="text"
                value={valueInput}
            />
        </div>
    );
}
