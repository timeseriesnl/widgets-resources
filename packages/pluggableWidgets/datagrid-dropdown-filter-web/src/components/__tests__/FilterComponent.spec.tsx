import { mount, shallow } from "enzyme";
import { createElement } from "react";
import { FilterComponent } from "../FilterComponent";
import { ListAttributeValueBuilder } from "@mendix/piw-utils-internal";

const defaultOptions = [
    { caption: "1", value: "_1" },
    { caption: "2", value: "_2" },
    { caption: "3", value: "_3" }
];

describe("Filter selector", () => {
    describe("with single selection", () => {
        describe("renders correctly", () => {
            it("with options", () => {
                const component = shallow(<FilterComponent options={defaultOptions} />);

                expect(component).toMatchSnapshot();
            });
            it("with no options", () => {
                const component = shallow(<FilterComponent options={[]} />);

                expect(component).toMatchSnapshot();
            });
            it("with ariaLabel", () => {
                const component = shallow(<FilterComponent options={defaultOptions} ariaLabel="my label" />);

                expect(component).toMatchSnapshot();
            });
            it("with emptyOptioncaption", () => {
                const component = shallow(<FilterComponent options={defaultOptions} emptyOptionCaption={"find me"} />);

                expect(component).toMatchSnapshot();
                expect(component.find("input").first().prop("placeholder")).toBe("find me");
            });

            it("with automatic options from the attribute", () => {
                const component = mount(
                    <FilterComponent
                        auto
                        attribute={new ListAttributeValueBuilder()
                            .withUniverse(["enum_value_1", "enum_value_2"])
                            .build()}
                        options={[]}
                    />
                );

                const input = component.find("input.dropdown-triggerer");
                input.simulate("click");

                expect(component).toMatchSnapshot();
            });
        });
        it("selects default option", () => {
            const defaultOption = defaultOptions[0];
            const component = shallow(<FilterComponent options={defaultOptions} defaultValue={defaultOption.value} />);

            const input = component.find("input").first();

            expect(input.prop("value")).toBe(defaultOption.caption);
        });

        describe("when value changes", () => {
            it("calls updateFilters when value changes", () => {
                const updateFilterHandler = jest.fn();
                const component = shallow(
                    <FilterComponent options={defaultOptions} updateFilters={updateFilterHandler} />
                );

                const input = component.find("input");
                input.simulate("click");

                const item = component.find("li").first();
                item.simulate("click");

                expect(updateFilterHandler).toBeCalled();
            });
            it("shows selected option on input value", () => {
                const defaultOption = defaultOptions[1];
                const component = shallow(
                    <FilterComponent options={defaultOptions} defaultValue={defaultOption.value} />
                );

                const input = component.find("input");
                input.simulate("click");

                expect(component.find("input").first().prop("value")).toBe(defaultOption.caption);

                const item = component.find("li").last(); // [cap 3: val:_3]
                item.simulate("click");

                expect(component.find("input").first().prop("value")).toBe(defaultOptions[2].caption);
            });
        });
    });

    describe("with multi selection", () => {
        describe("renders correctly", () => {
            it("with options", () => {
                const component = shallow(<FilterComponent multiSelect options={defaultOptions} />);

                expect(component).toMatchSnapshot();
            });
            it("with no options", () => {
                const component = shallow(<FilterComponent multiSelect options={[]} />);

                expect(component).toMatchSnapshot();
            });
            it("with ariaLabel", () => {
                const component = shallow(
                    <FilterComponent options={defaultOptions} multiSelect ariaLabel="my label" />
                );

                expect(component).toMatchSnapshot();
            });
            it("with emptyOptioncaption", () => {
                const component = shallow(
                    <FilterComponent multiSelect options={defaultOptions} emptyOptionCaption={"find me"} />
                );

                expect(component).toMatchSnapshot();
                expect(component.find("input").first().prop("placeholder")).toBe("find me");
            });
        });

        describe("with default options set", () => {
            it("selects single default option", () => {
                const defaultOption = defaultOptions[0];
                const component = shallow(
                    <FilterComponent multiSelect options={defaultOptions} defaultValue={defaultOption.value} />
                );

                const input = component.find("input").first();

                expect(input.prop("value")).toBe(defaultOption.caption);
            });

            it("selects multiple default options", () => {
                const defaultValue = `${defaultOptions[0].value},${defaultOptions[1].value}`;

                const component = shallow(
                    <FilterComponent multiSelect options={defaultOptions} defaultValue={defaultValue} />
                );

                const input = component.find("input").first();
                const expectedCaptions = `${defaultOptions[0].caption},${defaultOptions[1].caption}`;
                expect(input.prop("value")).toBe(expectedCaptions);
            });

            it("filters incorrect default options", () => {
                const inCorrectDefaultValue = `${defaultOptions[0].value},${defaultOptions[1].value},SomeRandomText`;

                const component = shallow(
                    <FilterComponent multiSelect options={defaultOptions} defaultValue={inCorrectDefaultValue} />
                );

                const input = component.find("input").first();

                const expectedCaptions = `${defaultOptions[0].caption},${defaultOptions[1].caption}`;
                expect(input.prop("value")).toBe(expectedCaptions);
            });
        });

        describe("when value changes", () => {
            it("calls updateFilters when value changes", () => {
                const updateFiltersHandler = jest.fn();
                const component = shallow(
                    <FilterComponent multiSelect options={defaultOptions} updateFilters={updateFiltersHandler} />
                );

                const input = component.find("input");
                input.simulate("click");

                const item = component.find("li").first();
                item.simulate("click");

                expect(updateFiltersHandler).toBeCalled();
            });
            it("shows selected options on input value", () => {
                const component = shallow(<FilterComponent multiSelect options={defaultOptions} />);

                const input = component.find("input");
                input.simulate("click");

                const item = component.find("li").at(1);
                item.simulate("click");
                const item2 = component.find("li").at(2);
                item2.simulate("click");

                expect(component.find("input").first().prop("value")).toBe("2,3");
            });
        });
    });
});
