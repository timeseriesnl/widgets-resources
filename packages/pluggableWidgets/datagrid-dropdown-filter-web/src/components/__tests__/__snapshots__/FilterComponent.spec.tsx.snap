// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter selector with multi selection renders correctly with ariaLabel 1`] = `
<div
  className="dropdown-container"
  data-focusindex={0}
>
  <input
    aria-controls="undefined-dropdown-list"
    aria-expanded={false}
    aria-haspopup={true}
    aria-label="my label"
    className="form-control dropdown-triggerer"
    onClick={[Function]}
    onFocus={[Function]}
    value=""
  />
</div>
`;

exports[`Filter selector with multi selection renders correctly with emptyOptioncaption 1`] = `
<div
  className="dropdown-container"
  data-focusindex={0}
>
  <input
    aria-controls="undefined-dropdown-list"
    aria-expanded={false}
    aria-haspopup={true}
    className="form-control dropdown-triggerer"
    onClick={[Function]}
    onFocus={[Function]}
    placeholder="find me"
    value=""
  />
</div>
`;

exports[`Filter selector with multi selection renders correctly with no options 1`] = `
<div
  className="dropdown-container"
  data-focusindex={0}
>
  <input
    aria-controls="undefined-dropdown-list"
    aria-expanded={false}
    aria-haspopup={true}
    className="form-control dropdown-triggerer"
    onClick={[Function]}
    onFocus={[Function]}
    value=""
  />
</div>
`;

exports[`Filter selector with multi selection renders correctly with options 1`] = `
<div
  className="dropdown-container"
  data-focusindex={0}
>
  <input
    aria-controls="undefined-dropdown-list"
    aria-expanded={false}
    aria-haspopup={true}
    className="form-control dropdown-triggerer"
    onClick={[Function]}
    onFocus={[Function]}
    value=""
  />
</div>
`;

exports[`Filter selector with single selection renders correctly with ariaLabel 1`] = `
<div
  className="dropdown-container"
  data-focusindex={0}
>
  <input
    aria-controls="undefined-dropdown-list"
    aria-expanded={false}
    aria-haspopup={true}
    aria-label="my label"
    className="form-control dropdown-triggerer"
    onClick={[Function]}
    onFocus={[Function]}
    value=""
  />
</div>
`;

exports[`Filter selector with single selection renders correctly with automatic options from the attribute 1`] = `
<FilterComponent
  attribute={
    Object {
      "filterable": true,
      "formatter": Object {
        "format": [MockFunction] {
          "calls": Array [
            Array [
              "enum_value_1",
            ],
            Array [
              "enum_value_2",
            ],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": "enum_value_1",
            },
            Object {
              "type": "return",
              "value": "enum_value_2",
            },
          ],
        },
        "parse": [MockFunction],
      },
      "get": [MockFunction],
      "id": "",
      "sortable": true,
      "type": "String",
      "universe": Array [
        "enum_value_1",
        "enum_value_2",
      ],
    }
  }
  auto={true}
  options={Array []}
>
  <div
    className="dropdown-container"
    data-focusindex={0}
  >
    <input
      aria-controls="undefined-dropdown-list"
      aria-expanded={true}
      aria-haspopup={true}
      className="form-control dropdown-triggerer"
      onClick={[Function]}
      onFocus={[Function]}
      value=""
    />
    <ul
      className="dropdown-list"
      data-focusindex={0}
      id="undefined-dropdown-list"
      role="menu"
      style={
        Object {
          "width": 0,
        }
      }
    >
      <li
        className="filter-selected"
        key="0"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="menuitem"
        tabIndex={0}
      >
        <div
          className="filter-label"
        />
      </li>
      <li
        className=""
        key="1"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="menuitem"
        tabIndex={0}
      >
        <div
          className="filter-label"
        >
          enum_value_1
        </div>
      </li>
      <li
        className=""
        key="2"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="menuitem"
        tabIndex={0}
      >
        <div
          className="filter-label"
        >
          enum_value_2
        </div>
      </li>
    </ul>
  </div>
</FilterComponent>
`;

exports[`Filter selector with single selection renders correctly with emptyOptioncaption 1`] = `
<div
  className="dropdown-container"
  data-focusindex={0}
>
  <input
    aria-controls="undefined-dropdown-list"
    aria-expanded={false}
    aria-haspopup={true}
    className="form-control dropdown-triggerer"
    onClick={[Function]}
    onFocus={[Function]}
    placeholder="find me"
    value=""
  />
</div>
`;

exports[`Filter selector with single selection renders correctly with no options 1`] = `
<div
  className="dropdown-container"
  data-focusindex={0}
>
  <input
    aria-controls="undefined-dropdown-list"
    aria-expanded={false}
    aria-haspopup={true}
    className="form-control dropdown-triggerer"
    onClick={[Function]}
    onFocus={[Function]}
    value=""
  />
</div>
`;

exports[`Filter selector with single selection renders correctly with options 1`] = `
<div
  className="dropdown-container"
  data-focusindex={0}
>
  <input
    aria-controls="undefined-dropdown-list"
    aria-expanded={false}
    aria-haspopup={true}
    className="form-control dropdown-triggerer"
    onClick={[Function]}
    onFocus={[Function]}
    value=""
  />
</div>
`;
